README.md
# jellysub
A shim to use Subsonic clients with Jellyfin

Jellysub is an HTTP service which sits between Subsonic-compatible clients and a
Jellyfin server.

## Compatibility
Clients tested with Jellysub:

| Client                                          | Notes                                |
|-------------------------------------------------|--------------------------------------|
| [Audinaut](https://github.com/nvllsvm/audinaut) | fully functional excluding playlists |


## Installation
Available in the following distribution channels:
- [Docker](https://hub.docker.com/r/nvllsvm/jellysub)
- [PyPi (Python)](https://pypi.org/project/jellysub/)

## Running
The following environment variables may be used to configure Jellysub.

| Name                    | Description                                           |
|-------------------------|-------------------------------------------------------|
| `JELLYSUB_HTTP_PORT`    | Port to listen for HTTP requests on. (Default `4040`) |
| `JELLYSUB_UPSTREAM_URL` | URL of the Jellyfin server. Required.                 |

Command line arguments may also be used. See `jellysub --help` for more information.

### Examples
Both examples below start a Jellysub server on port **4040** and communicating
with the Jellyfin server at **https://yourjellyfinserver.com**.

#### Docker
```
docker run \
    -e JELLYSUB_UPSTREAM_URL=https://yourjellyfinserver.com \
    -p 4040:4040 \
    nvllsvm/jellysub
```

#### Command Line
```
jellysub --upstream https://yourjellyfinserver.com
```


CHANGELOG.md
# Changelog

## 0.0.6
- Fix compatibility with Jellyfin 10.8.0
- Fix compatibility with latest aiohttp

## 0.0.5
- Fix albums without production year

## 0.0.4
- Add `getGenres`
- Add support for either query string or POST bodies
- Add support for hex-encoded passwords
- Fix `getArtistInfo2`

## 0.0.3
- Add support for json responses (`f=json`)
- Add `getArtistInfo2` endpoint
- Fix song file suffix
- Fix missing artist name in `getArtist`
- Many other updates. This is a refactor heavy release - as to be expected for alpha-level software.

## 0.0.2
- Add album art

## 0.0.1
- Misc fixes

## 0.0.0
- Initial Release


tests/__init__.py
#!/usr/bin/env python3
import functools
import os
import pathlib
import subprocess

import aiohttp
import yarl

_TEST_DIR = pathlib.Path(__file__).parent


os.environ['COMPOSE_FILE'] = str(_TEST_DIR.joinpath('docker-compose.yml'))


@functools.lru_cache()
def bootstrap():
    subprocess.run(
        ['docker-compose', 'pull'],
        check=True)

    subprocess.run(
        ['docker-compose', 'down', '--timeout=0',
         '--volumes', '--remove-orphans'],
        check=True)

    subprocess.run(
        ['docker-compose', 'up', '-d'],
        check=True)

    return {
        'mockserver_url': yarl.URL(_docker_compose_port('mockserver', 1080))
    }


def _docker_compose_port(service, port):
    proc = subprocess.run(
        ['docker-compose', 'port', service, str(port)],
        stdout=subprocess.PIPE,
        check=True)
    return proc.stdout.decode()


async def mockserver_expectation(**kwargs):
    response = await aiohttp.put(
        bootstrap()['mockserver_url'] / 'mockserver/expectation',
        json=kwargs)


async def mockserver_reset():
    response = await aiohttp.put(
        bootstrap()['mockserver_url'] / 'mockserver/reset')


tests/test_app.py
import unittest

import aiohttp.test_utils
import aiohttp.web
import jellysub.app
import yarl


class MockJellyfinServer(aiohttp.web.Application):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.add_routes([
            aiohttp.web.route(
                'POST', '/Users/authenticatebyname', self.authenticate),
        ])

    @staticmethod
    async def authenticate(request):
        body = await request.json()
        if body['Username'] != request.app['username']:
            return aiohttp.web.Response(status=401)
        if body['Pw'] != request.app['password']:
            return aiohttp.web.Response(status=401)
        return aiohttp.web.json_response({'AccessToken': 'ZZZ'})


class AuthTests(unittest.IsolatedAsyncioTestCase):

    async def asyncSetUp(self):
        await super().asyncSetUp()

        self.username = 'abc'
        self.password = 'xyz'

        self.mock_jellyfin = aiohttp.test_utils.TestServer(
            MockJellyfinServer())

        self.mock_jellyfin.app['username'] = 'abc'
        self.mock_jellyfin.app['password'] = 'xyz'

        await self.mock_jellyfin.start_server()

        self.app = jellysub.app.Application(
            str(yarl.URL.build(
                scheme='http',
                host=self.mock_jellyfin.host,
                port=self.mock_jellyfin.port)))
        self.server = aiohttp.test_utils.TestServer(self.app)
        self.client = aiohttp.test_utils.TestClient(self.server)
        await self.client.start_server()
        self.addAsyncCleanup(self.client.close)
        self.addAsyncCleanup(self.app.cleanup)

    async def test_valid_auth(self):
        params = {
            'u': 'abc',
            'p': 'xyz'
        }
        resp = await self.client.request(
            'GET', '/rest/ping.view', params=params)
        self.assertEqual(resp.status, 200)

    async def test_invalid_auth(self):
        correct_params = {
            'u': self.username,
            'p': self.password,
        }

        for key in correct_params:
            params = correct_params.copy()
            params[key] = 'X'
            resp = await self.client.request(
                'GET', '/rest/ping.view', params=params)
            self.assertEqual(resp.status, 401)

    async def test_missing_required_query_param(self):
        required = {
            'u': self.username,
            'p': self.password,
        }
        for key in required:
            params = required.copy()
            params.pop(key)
            resp = await self.client.request(
                'GET', '/rest/ping.view', params=params)
            self.assertEqual(resp.status, 400)


jellysub/jellyfin.py
import importlib.metadata
import re
import uuid

import aiohttp.web
import yarl

try:
    _VERSION = importlib.metadata.version(__package__)
except importlib.metadata.PackageNotFoundError:
    _VERSION = 'unknown'


class JellyfinClient:
    _CLIENT = 'jellysub'
    _DEVICE = 'jellysub'
    _DEVICE_ID = uuid.uuid4().hex
    _VERSION = _VERSION

    def __init__(self, url):
        self._url = yarl.URL(url)
        self._users = {}
        self._auth_header = {
            'Client': self._CLIENT,
            'Device': self._DEVICE,
            'DeviceId': self._DEVICE_ID,
            'Version': self._VERSION,
        }

    async def open(self):
        self._client = aiohttp.ClientSession()

    async def close(self):
        await self._client.close()

    async def get_user(self, username, password):
        key = (username, password)
        if key not in self._users:
            user = await self._authenticate(username, password)
            if user:
                self._users[key] = user
        return self._users[key]

    async def get_album_artists(self, user):
        url = self._url / 'Artists/AlbumArtists'
        auth = dict(self._auth_header, Token=user['AccessToken'])
        kwargs = {
            'headers': {
                'X-Emby-Authorization': self._build_authorization_header(auth),
            },
            'params': {
                'Recursive': 'true',
                'StartIndex': '0',
                'Limit': '10000',
            }
        }
        async with self._client.get(url, **kwargs) as resp:
            return await resp.json()

    async def get_genres(self, user):
        url = self._url / 'Genres'
        auth = dict(self._auth_header, Token=user['AccessToken'])
        kwargs = {
            'headers': {
                'X-Emby-Authorization': self._build_authorization_header(auth),
            },
            'params': {
                'Recursive': 'true',
                'StartIndex': '0',
                'Limit': '10000',
            }
        }
        async with self._client.get(url, **kwargs) as resp:
            return await resp.json()

    async def get_artist(self, user, artist_id):
        url = self._url / 'Users' / user['User']['Id'] / 'Items' / artist_id
        auth = dict(self._auth_header, Token=user['AccessToken'])
        kwargs = {
            'headers': {
                'X-Emby-Authorization': self._build_authorization_header(auth),
            },
        }
        async with self._client.get(url, **kwargs) as resp:
            return await resp.json()

    async def get_albums(self, user, artist_id=None):
        url = self._url / 'Users' / user['User']['Id'] / 'Items'
        auth = dict(self._auth_header, Token=user['AccessToken'])

        params = {
            'IncludeItemTypes': 'MusicAlbum',
            'Limit': '1000',
            'Recursive': 'true',
            'StartIndex': '0',
        }

        if artist_id is not None:
            params['AlbumArtistIds'] = artist_id

        kwargs = {
            'headers': {
                'X-Emby-Authorization': self._build_authorization_header(auth),
            },
            'params': params
        }
        async with self._client.get(url, **kwargs) as resp:
            return await resp.json()

    async def get_album(self, user, album_id):
        url = self._url / 'Users' / user['User']['Id'] / 'Items'

        auth = dict(self._auth_header, Token=user['AccessToken'])
        kwargs = {
            'headers': {
                'X-Emby-Authorization': self._build_authorization_header(auth),
            },
            'params': {
                'ParentId': album_id,
                'Fields': 'MediaSources',
            }
        }
        async with self._client.get(url, **kwargs) as resp:
            return await resp.json()

    async def download_song(self, user, song_id):
        url = self._url / 'Items' / song_id / 'Download'

        kwargs = {
            'params': {
                'api_key': user['AccessToken']
            }
        }
        async with self._client.get(url, **kwargs) as resp:
            return await resp.read()

    async def get_album_cover(self, album_id):
        url = self._url / 'Items' / album_id / 'Images/Primary/0'

        kwargs = {
            'params': {
                'quality': '90'
            }
        }
        async with self._client.get(url, **kwargs) as resp:
            if resp.status != 200:
                raise ValueError
            return await resp.read()

    async def _authenticate(self, username, password):
        kwargs = {
            'allow_redirects': False,
            'headers': {
                'X-Emby-Authorization': self._build_authorization_header(
                    self._auth_header),
            },
            'json': {
                'Username': username,
                'Pw': password,
            },
        }
        url = self._url / 'Users/authenticatebyname'
        async with self._client.post(url, **kwargs) as resp:
            if resp.status == 200:
                return await resp.json()

    @staticmethod
    def _parse_authorization_header(header):
        # these are always required
        result = {
            'Device': None,
            'DeviceId': None,
            'Version': None,
        }
        # this is only required when accessing authorized resource
        result['Token'] = None

        # required during initial auth
        result['Client'] = None

        AUTH_REGEX = re.compile(r'\s(\w+)="(\S*)"')

        for x in AUTH_REGEX.finditer(header):
            key, value = x.groups()
            if key not in result or result[key] is not None:
                raise ValueError(header)
            result[key] = value

        if not all([v
                    for k, v in result.items()
                    if k not in ('Client', 'Token')]):
            raise ValueError(header)
        return result

    @staticmethod
    def _build_authorization_header(data):
        return 'MediaBrowser {}'.format(
            ', '.join(f'{key}="{value}"' for key, value in data.items()))


jellysub/app.py
import asyncio
import collections
import json
import string
from xml.etree.ElementTree import Element, tostring

import aiohttp.web


from . import jellyfin


@aiohttp.web.middleware
async def request_data_middleware(request, handler):
    data = {}
    for key, value in request.url.query.items():
        data[key] = value
    post_body = await request.post()
    for key, value in post_body.items():
        data[key] = value
    request['data'] = data
    return await handler(request)


@aiohttp.web.middleware
async def auth_middleware(request, handler):
    try:
        username = request['data']['u']
        password = request['data']['p']
    except KeyError:
        return aiohttp.web.Response(status=400)

    if password.startswith('enc:'):
        password = bytearray.fromhex(password.split(':')[1]).decode()

    try:
        request.user = await request.app['jellyfin'].get_user(
            username, password)
    except KeyError:
        return aiohttp.web.Response(status=401)
    return await handler(request)


@aiohttp.web.middleware
async def content_format_middleware(request, handler):
    response_format = request['data'].get('f', 'xml')
    if response_format not in ('json', 'xml'):
        return aiohttp.web.Response(status=400)

    response = await handler(request)

    content = response.pop('content', None)
    if content is not None:
        content.setdefault('status', 'ok')
        content.setdefault('version', '1.9.0')
        content = {'subsonic-response': content}

        if response_format == 'xml':
            response.body = tostring(_to_xml(content))
            response.content_type = 'application/xml'
        elif response_format == 'json':
            response.body = json.dumps(content)
            response.content_type = 'application/json'
    return response


def _to_xml(data, key=None):
    if key is None:
        if not isinstance(data, dict) or len(data) != 1:
            raise ValueError
        key = list(data.keys())[0]
        data = data[key]

    root = Element(key)
    for key, value in data.items():
        if isinstance(value, dict):
            root.append(_to_xml(value, key))
        elif isinstance(value, str):
            root.set(key, value)
        elif isinstance(value, (float, int, bool)):
            root.set(key, str(value))
        elif isinstance(value, list):
            root.extend([_to_xml(v, key) for v in value])
        elif value is None:
            root.set(key, '')
        else:
            raise ValueError
    return root


async def ping(request):
    response = aiohttp.web.Response()
    response['content'] = {}
    return response


async def artists(request):
    data = await request.app['jellyfin'].get_album_artists(request.user)

    results = collections.defaultdict(list)
    for item in data['Items']:
        first = item['Name'].lower()
        group = first if first in string.ascii_lowercase else '#'

        results[group].append(item)

    indexes = []
    for prefix, artists in results.items():
        artists = [
            {
                'albumCount': 1,  # TODO: determine real album count
                'id': artist['Id'],
                'name': artist['Name'],
            }
            for artist in artists
        ]
        artists = sorted(
            artists, key=lambda s: (s['name'], s['id']))

        indexes.append({
            'name': prefix,
            'artist': artists
        })
    indexes = sorted(indexes, key=lambda s: s['name'])

    response = aiohttp.web.Response()
    response['content'] = {
        'artists': {
            'ignoredArticles': '',
            'index': indexes
        },
    }
    return response


async def genres(request):
    data = await request.app['jellyfin'].get_genres(request.user)

    genres = [
        {
            'value': item['Name'],

            # TODO: look these up
            'albumCount': 1,
            'songCount': 1,
        }
        for item in data['Items']
    ]

    genres = sorted(genres, key=lambda s: s['value'])

    response = aiohttp.web.Response()
    response['content'] = {
        'genres': {
            'genre': genres
        }
    }
    return response


async def artist(request):
    artist_id = request['data']['id']
    artist_data, album_data = await asyncio.gather(
        request.app['jellyfin'].get_artist(request.user, artist_id),
        request.app['jellyfin'].get_albums(request.user, artist_id)
    )

    # TODO can AlbumArtist and Name be null?
    albums = [
        {
            'artist': item['AlbumArtist'],
            'artistId': artist_id,
            'coverArt': item['Id'],
            'id': item['Id'],
            'name': item['Name'],
            'duration': 0,
            'songCount': 0,
            'year': item.get('ProductionYear'),
        }
        for item in album_data['Items']
    ]
    albums = sorted(albums, key=lambda s: (s['year'] or 0, s['name'], s['id']))

    response = aiohttp.web.Response()
    response['content'] = {
        'artist': {
            'album': albums,
            'albumCount': len(albums),
            'id': artist_id,
            'name': artist_data['Name'],
        },
    }
    return response


async def artist_info2(request):
    artist_id = request['data']['id']
    data = await request.app['jellyfin'].get_artist(request.user, artist_id)

    response = aiohttp.web.Response()
    response['content'] = {
        'artistInfo2': {
            'biography': data.get('Overview', ''),
        },
    }
    return response


async def album(request):
    album_id = request['data']['id']
    data = await request.app['jellyfin'].get_album(request.user, album_id)

    songs = []
    for item in data['Items']:
        path = item['MediaSources'][0]['Path']
        suffix = path.split('.')[-1] if '.' in path else ''
        song = {
            'id': item['Id'],
            'artist': ' & '.join(item['Artists']),
            'album': item['Album'],
            'title': item['Name'],
            'coverArt': item['AlbumId'],
            'duration': int(item['RunTimeTicks'] / 10000000),
            'track': item['IndexNumber'],
            'path': path,
            'suffix': suffix,
        }
        songs.append(song)
    songs = sorted(songs, key=lambda s: (s['track'], s['title'], s['id']))

    response = aiohttp.web.Response()
    response['content'] = {
        'album': {
            'song': songs,
            'songCount': len(songs),
            'id': album_id
        },
    }
    return response


async def albumList(request):
    offset = int(request['data'].get('offset', 0))
    size = int(request['data'].get('size', 10))

    data = await request.app['jellyfin'].get_albums(request.user)

    albums = []
    for (i, item) in enumerate(data['Items']):
        if i < offset:
            continue
        if (i - offset) == size:
            break

        album_data = await request.app['jellyfin'].get_album(request.user, item['Id'])

        albums.append({
            'id': item['Id'],
            'parent': 1,
            'title': item['Name'],
            'artist': item['AlbumArtists'][0]['Name'],
            'isDir': True,
            'coverArt': item['Id'],
        })

    response = aiohttp.web.Response()
    response['content'] = {
        'albumList': {
            'album': albums
        },
    }
    return response

async def albumList2(request):
    offset = int(request['data'].get('offset', 0))
    size = int(request['data'].get('size', 10))

    data = await request.app['jellyfin'].get_albums(request.user)

    albums = []
    for (i, item) in enumerate(data['Items']):
        if i < offset:
            continue
        if (i - offset) == size:
            break

        album_data = await request.app['jellyfin'].get_album(request.user, item['Id'])

        albums.append({
            'id': item['Id'],
            'name': item['Name'],
            'coverArt': item['Id'],
            'songCount': len(album_data['Items']),
            'artist': item['AlbumArtists'][0]['Name'],
            'artistId': item['AlbumArtists'][0]['Id'],
            'duration': int(item['RunTimeTicks'] / 10000000)
        })

    response = aiohttp.web.Response()
    response['content'] = {
        'albumList2': {
            'album': albums
        },
    }
    return response


async def cover_art(request):
    album_id = request['data']['id']
    data = await request.app['jellyfin'].get_album_cover(album_id)

    return aiohttp.web.Response(body=data)


async def stream(request):
    song_id = request['data']['id']
    data = await request.app['jellyfin'].download_song(request.user, song_id)

    return aiohttp.web.Response(body=data)


class Application(aiohttp.web.Application):

    def __init__(self, upstream):
        super().__init__(
            middlewares=[
                request_data_middleware,
                content_format_middleware,
                auth_middleware
            ])

        self['jellyfin'] = jellyfin.JellyfinClient(upstream)

        self.add_routes([
            aiohttp.web.route('*', '/rest/ping.view', ping),
            aiohttp.web.route('*', '/rest/getArtists.view', artists),
            aiohttp.web.route('*', '/rest/getArtist.view', artist),
            aiohttp.web.route('*', '/rest/getAlbum.view', album),
            aiohttp.web.route('*', '/rest/getAlbumList.view', albumList),
            aiohttp.web.route('*', '/rest/getAlbumList2.view', albumList2),
            aiohttp.web.route('*', '/rest/stream.view', stream),
            aiohttp.web.route('*', '/rest/getCoverArt.view', cover_art),
            aiohttp.web.route('*', '/rest/getArtistInfo2.view', artist_info2),
            aiohttp.web.route('*', '/rest/getGenres.view', genres),
        ])
        self.on_startup.append(self.on_start)

    async def on_start(self, *_):
        await self['jellyfin'].open()

    async def cleanup(self):
        await super().cleanup()
        await self['jellyfin'].close()
